src/extra/utils/functions.py

```diff
@@ -23,6 +23,53 @@ def gaussian(x, y0, A, μ, σ):
     return y0 + (A / (σ * np.sqrt(2*np.pi))) * np.exp(-(x - μ)**2 / (2 * σ**2))
 
 
+def fit_gaussian(ydata, return_all=False, xdata=None, p0=None, **kwargs):
+    """Fit a Gaussian to some data.
+
+    This uses [curve_fit()][scipy.optimize.curve_fit] to fit a Gaussian (from
+    [gaussian()][extra.utils.gaussian]) to `ydata`. If `p0` is not passed the
+    function will set them to reasonable defaults. It will return `None` if
+    fitting fails, or if there are no non-NaN values in `ydata`.
+
+    Args:
+        ydata (array_like): The data to fit. NaN's will automatically be masked
+            before fitting.
+        return_all (bool): Whether to return everything from
+            [curve_fit()][scipy.optimize.curve_fit], or just `popt`.
+        p0 (list): A list of `[y0, A, μ, σ]` to match the arguments to
+            [gaussian()][extra.utils.gaussian].
+        **kwargs (): All other keyword arguments will be passed to
+            [curve_fit()][scipy.optimize.curve_fit].
+    """
+    if xdata is None:
+        xdata = np.arange(len(ydata))
+
+    from scipy.optimize import curve_fit
+
+    # Filter nans
+    nan_mask = ~np.isnan(ydata)
+    xdata = xdata[nan_mask]
+    ydata = ydata[nan_mask]
+
+    if len(ydata) == 0:
+        return None
+
+    if p0 is None:
+        μ_idx = np.argmax(ydata)
+
+        μ = xdata[μ_idx]
+        A = max(1, ydata[μ_idx])
+        σ = abs(np.max(xdata) - np.min(xdata)) / 4
+        y0 = np.min(ydata)
+        p0 = [y0, A, μ, σ]
+
+    try:
+        result = curve_fit(gaussian, xdata, ydata, p0=p0)
+        return result if return_all else result[0]
+    except RuntimeError:
+        return None
```

Comment:
It seems like it would be easy to assume this function succeeds, do something with the results like `y0, A, μ, σ = fit_gaussian(...)` and get a confusing TypeError if the fit fails.

Unless there's a good reason why returning None is helpful, I think it's probably better to just let the RuntimeError propagate.

Reply:
I actually prefer returning `None` because my usual use-case for this is fitting entire runs of data per-train where it's normal for the fitting to occasionally fail, and I'd rather this return `None` than having to remember the right exception to catch every time.

Reply:
OK, fair enough.